ext/iodine/iodine_http.c

```diff
@@ -365,7 +365,11 @@ static inline VALUE copy2env(iodine_http_request_handle_s *handle) {
   if (*pos == 0) {
     rb_hash_aset(env, SERVER_NAME,
                  rb_enc_str_new(tmp.data, tmp.len, IodineBinaryEncoding));
-    rb_hash_aset(env, SERVER_PORT, QUERY_ESTRING);
+    if (RSTRING_LEN(QUERY_ESTRING) != 0) {
+      rb_hash_aset(env, SERVER_PORT, QUERY_ESTRING);
+    } else {
+      rb_hash_delete(env, SERVER_PORT);
```

Comment:
I think that using `rb_hash_aset(env, SERVER_PORT, Qnil)` would be both faster here and in the client code.

Using the `delete` operator can cause possible hash map misses when searching for a value and these misses are slower than finding a `nil` value.

This should work with the way the Linter middleware checks the value: https://github.com/rack/rack/blob/d15dd728440710cfc35ed155d66a98dc2c07ae42/lib/rack/lint.rb#L284

Reply:
Unfortunately I tried `Qnil` originally, and it failed. I just tried it again to confirm.

`ERROR: Iodine caught an unprotected exception - Rack::Lint::LintError: env variable SERVER_PORT has non-string value nil`

```diff
diff --git a/ext/iodine/iodine_http.c b/ext/iodine/iodine_http.c
index 4f7475f..a9ca702 100644
--- a/ext/iodine/iodine_http.c
+++ b/ext/iodine/iodine_http.c
@@ -365,11 +365,7 @@ static inline VALUE copy2env(iodine_http_request_handle_s *handle) {
   if (*pos == 0) {
     rb_hash_aset(env, SERVER_NAME,
                  rb_enc_str_new(tmp.data, tmp.len, IodineBinaryEncoding));
-    if (RSTRING_LEN(QUERY_ESTRING) != 0) {
-      rb_hash_aset(env, SERVER_PORT, QUERY_ESTRING);
-    } else {
-      rb_hash_delete(env, SERVER_PORT);
-    }
+    rb_hash_aset(env, SERVER_PORT, Qnil);
   } else {
     rb_hash_aset(
         env, SERVER_NAME,
```

This is because the described spec is wrong, and non-String values are failures too, if they don't contain a period: https://github.com/rack/rack/blob/d15dd728440710cfc35ed155d66a98dc2c07ae42/lib/rack/lint.rb#L308-L315

Reply:
Actually, I think the spec is right, they are just using the presence of `nil` to guard on the coercion through `Integer`. So it looks like it allows `nil`, but in essence they only care in that scenario that they don't coerce a `nil` value and trigger an error at that line.

My guess anyways.

Reply:
I could drop the conditional and always delete. Or could we remove the empty string that is initialized earlier in this file (`rb_hash_aset(env_template_no_upgrade, SERVER_PORT, QUERY_ESTRING);`)?

```
diff --git a/ext/iodine/iodine_http.c b/ext/iodine/iodine_http.c
index 4f7475f..7f47c17 100644
--- a/ext/iodine/iodine_http.c
+++ b/ext/iodine/iodine_http.c
@@ -365,11 +365,7 @@ static inline VALUE copy2env(iodine_http_request_handle_s *handle) {
   if (*pos == 0) {
     rb_hash_aset(env, SERVER_NAME,
                  rb_enc_str_new(tmp.data, tmp.len, IodineBinaryEncoding));
-    if (RSTRING_LEN(QUERY_ESTRING) != 0) {
-      rb_hash_aset(env, SERVER_PORT, QUERY_ESTRING);
-    } else {
-      rb_hash_delete(env, SERVER_PORT);
-    }
+    rb_hash_delete(env, SERVER_PORT);
   } else {
     rb_hash_aset(
         env, SERVER_NAME,
```